pyquil/api/_config.py

```diff
@@ -93,65 +135,164 @@ class PyquilConfig(object):
         "file": FOREST_CONFIG,
         "section": "Rigetti Forest",
         "name": "qpu_compiler_address",
+        "engagement_key": 'qpu_compiler_endpoint',
         "default": None
     }
 
-    def __init__(self):
-        self.configparsers = {}
-        for env_name, default_path in CONFIG_PATHS.items():
+    def __init__(self, config_paths: Dict[str, str] = CONFIG_PATHS):
+        """
+        :param config_paths: the paths to the various configuration files read by PyQuil
+        """
+
+        # The engagement callback can be added by config consumers after construction
+        self.get_engagement = lambda: None
+
+        self.config_parsers = {}
+        for env_name, default_path in config_paths.items():
             default_path = expanduser(default_path)
             path = environ.get(env_name, default_path)
 
             cp = ConfigParser()
             cp.read(abspath(path))
-            self.configparsers[env_name] = cp
+            self.config_parsers[env_name] = cp
+        self._parse_auth_tokens()
 
-    def _env_or_config_or_default(self, env=None, file=None, section=None, name=None, default=None):
+    def _parse_auth_tokens(self):
+        self.user_auth_token = _parse_auth_token(self.user_auth_token_path, ['access_token', 'refresh_token', 'scope'])
+        self.qmi_auth_token = _parse_auth_token(self.qmi_auth_token_path, ['access_token', 'refresh_token'])
+
+    def _env_or_config_or_default(self,
+                                  env=None,
+                                  file=None,
+                                  section=None,
+                                  name=None,
+                                  engagement_key=None,
```

Comment:
why this not last?

Reply:
Following the order of precedence and how they're stored in the dicts. Since this is a 'private' method, bw compatibility wasn't a concern